src/server/pkg/storage/fileset/merge.go

```diff
@@ -0,0 +1,202 @@
+package fileset
+
+import (
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/tar"
+)
+
+type stream interface {
+	next() error
+	key() string
+}
+
+type mergeFunc func([]stream) error
+
+type fileStream struct {
+	r   *Reader
+	hdr *index.Header
+}
+
+func (fs *fileStream) next() error {
+	var err error
+	fs.hdr, err = fs.r.Next()
+	return err
+}
+
+func (fs *fileStream) key() string {
+	return fs.hdr.Hdr.Name
+}
+
+func idxMergeFunc(w *Writer) mergeFunc {
+	return func(ss []stream) error {
+		// (bryce) this will implement an index merge, which will be used by the distributed merge process.
+		return nil
+	}
+}
+
+func contentMergeFunc(w *Writer) mergeFunc {
+	return func(ss []stream) error {
+		// Convert generic streams to file streams.
+		var fileStreams []*fileStream
+		for _, s := range ss {
+			fileStreams = append(fileStreams, s.(*fileStream))
+		}
+		// Setup tag streams for tag merge.
+		var tagStreams []stream
+		var size int64
+		for _, fs := range fileStreams {
+			tagStreams = append(tagStreams, &tagStream{
+				r: fs.r,
+				// (bryce) header tag removed by first next call.
+				tags: fs.hdr.Idx.DataOp.Tags,
+			})
+			size += fs.hdr.Idx.SizeBytes
+		}
+		// Write header for file.
+		hdr := &index.Header{
+			Hdr: &tar.Header{
+				Name: fileStreams[0].hdr.Hdr.Name,
+				Size: size,
+			},
+		}
+		if err := w.WriteHeader(hdr); err != nil {
+			return err
+		}
+		// Merge file content.
+		return merge(tagStreams, tagMergeFunc(w))
+	}
+}
+
+type tagStream struct {
+	r    *Reader
+	tags []*index.Tag
+}
+
+func (ts *tagStream) next() error {
+	ts.tags = ts.tags[1:]
+	if len(ts.tags) == 0 {
+		return io.EOF
+	}
+	return nil
+}
+
+func (ts *tagStream) key() string {
+	return ts.tags[0].Id
+}
+
+func tagMergeFunc(w *Writer) mergeFunc {
+	return func(ss []stream) error {
+		// (bryce) this should be an Internal error type.
+		if len(ss) > 1 {
+			return fmt.Errorf("tags should be distinct within a file")
+		}
+		// Convert generic stream to tag stream.
+		tagStream := ss[0].(*tagStream)
+		// Copy tagged data to writer.
+		w.StartTag(tagStream.tags[0].Id)
+		return CopyN(w, tagStream.r, tagStream.tags[0].SizeBytes)
+	}
+}
+
+type mergePriorityQueue struct {
+	queue []stream
+	f     mergeFunc
+	size  int
+}
```

Comment:
In general, this priority queue implementation seems a little strange - it doesn't encapsulate its properties very well as the `merge` function is very cognizant of the internal workings, and so the code is a bit hard to follow.  Overall not too bad, since its contained within this one file, but it might be nice to at least add a comment block about the invariants of the `mergePriorityQueue`.